components/suggest/src/db.rs

```diff
@@ -1407,6 +1210,133 @@ impl<'a> FullKeywordInserter<'a> {
+            "INSERT INTO suggestions(
+                 record_id,
+                 title,
+                 url,
+                 score,
+                 provider
+             )
+             VALUES(?, ?, ?, ?, ?)
+             RETURNING id",
+        )?))
+    }
+
+    /// Execute the insert and return the `suggestion_id` for the new row
+    fn execute(
+        &mut self,
+        record_id: &SuggestRecordId,
+        title: &str,
+        url: &str,
+        score: f64,
+        provider: SuggestionProvider,
+    ) -> Result<i64> {
+        Ok(self.0.query_row(
+            (record_id.as_str(), title, url, score, provider as u8),
+            |row| row.get(0),
+        )?)
+    }
+}
+
+struct AmpInsertStatement<'conn>(rusqlite::Statement<'conn>);
+
+impl<'conn> AmpInsertStatement<'conn> {
+    fn new(conn: &'conn Connection) -> Result<Self> {
+        Ok(Self(conn.prepare(
+            "INSERT INTO amp_custom_details(
+                 suggestion_id,
+                 advertiser,
+                 block_id,
+                 iab_category,
+                 impression_url,
+                 click_url,
+                 icon_id
+             )
+             VALUES(?, ?, ?, ?, ?, ?, ?)
+             ",
+        )?))
+    }
+
+    fn execute(&mut self, suggestion_id: i64, amp: &DownloadedAmpSuggestion) -> Result<()> {
+        self.0.execute((
+            suggestion_id,
+            &amp.advertiser,
+            amp.block_id,
+            &amp.iab_category,
+            &amp.impression_url,
+            &amp.click_url,
+            &amp.icon_id,
+        ))?;
+        Ok(())
+    }
+}
+
+struct WikipediaInsertStatement<'conn>(rusqlite::Statement<'conn>);
+
+impl<'conn> WikipediaInsertStatement<'conn> {
+    fn new(conn: &'conn Connection) -> Result<Self> {
+        Ok(Self(conn.prepare(
+            "INSERT INTO wikipedia_custom_details(
+                 suggestion_id,
+                 icon_id
+             )
+             VALUES(?, ?)
+             ",
+        )?))
+    }
+
+    fn execute(
+        &mut self,
+        suggestion_id: i64,
+        wikipedia: &DownloadedWikipediaSuggestion,
+    ) -> Result<()> {
+        self.0.execute((suggestion_id, &wikipedia.icon_id))?;
+        Ok(())
+    }
+}
+
+struct KeywordInsertStatement<'conn>(rusqlite::Statement<'conn>);
+
+impl<'conn> KeywordInsertStatement<'conn> {
+    fn new(conn: &'conn Connection) -> Result<Self> {
+        Ok(Self(conn.prepare(
+            "INSERT INTO keywords(
+                 suggestion_id,
+                 keyword,
+                 full_keyword_id,
+                 rank
+             )
+             VALUES(?, ?, ?, ?)
+             ",
+        )?))
+    }
+
+    fn execute(
+        &mut self,
+        suggestion_id: i64,
+        keyword: &str,
+        full_keyword_id: Option<i64>,
+        rank: usize,
+    ) -> Result<()> {
+        self.0
+            .execute((suggestion_id, keyword, full_keyword_id, rank))?;
+        Ok(())
+    }
+}
+
```

Comment:
We could define more of these statement types, for example to handle the inserting pocket/MDN/Weather data, or maybe even for the fetch/deletion code.  However, for this PR I only targetted the queries that I think would make a difference to the overall ingestion time for a fresh DB.

Reply:
Makes sense. Can you add a note for this to guide folks on when/how to adopt this treatment?